moquette-0.17/broker/src/main/java/io/moquette/broker/Server.java

```diff
@@ -181,52 +196,262 @@ public void startServer(IConfig config, List<? extends InterceptHandler> handler
+                "\"remote_ip\": \"%s\", " +
+                "\"uuid\": \"%s\"}",
+            os, cpuArch, jvmVersion, jvmVendor, MOQUETTE_VERSION, this.standalone, maxHeap, remoteIp, uuid);
+    }
+
+    private String retrievePublicIP() {
+        try {
+            URL url = new URL("http://whatismyip.akamai.com");
+            HttpURLConnection con = (HttpURLConnection) url.openConnection();
+            final int status = con.getResponseCode();
+            if (status != HttpURLConnection.HTTP_OK) {
+                LOG.debug("What's my IP service replied with {}", status);
+                return "";
+            }
+
+            BufferedReader in = new BufferedReader(new InputStreamReader(con.getInputStream()));
+            return in.readLine();
+        } catch (Exception e) {
+            LOG.debug("Can't connect to what's my IP service");
+            return "";
+        }
+    }
+
+    private void sendTelemetryData(String telemetryDoc) throws IOException {
+        URL url = new URL("https://telemetry.moquette.io/api/v1/notify");
+        HttpURLConnection con = (HttpURLConnection) url.openConnection();
+        con.setRequestMethod("POST");
+        con.setRequestProperty("Content-Type", "application/json");
+        con.setRequestProperty("Accept", "application/json");
+        con.setInstanceFollowRedirects(true);
+
+        // POST
+        con.setDoOutput(true);
+        final byte[] input = telemetryDoc.getBytes("utf-8");
+        try (OutputStream os = con.getOutputStream()) {
+            os.write(input, 0, input.length);
+        }
+
+        int status = con.getResponseCode();
+        LOG.trace("Response code is {}", status);
+
+        boolean redirect = false;
+
+        // normally, 3xx is redirect
+        if (status != HttpURLConnection.HTTP_OK) {
+            if (status == HttpURLConnection.HTTP_MOVED_TEMP
+                || status == HttpURLConnection.HTTP_MOVED_PERM
+                || status == HttpURLConnection.HTTP_SEE_OTHER)
+                redirect = true;
+        }
+
+        LOG.trace("Response Code: {} ", status);
+
+        if (redirect) {
+
+            // get redirect url from "location" header field
+            String newUrl = con.getHeaderField("Location");
+
+            // open the new connnection again
+            con = (HttpURLConnection) new URL(newUrl).openConnection();
+            con.addRequestProperty("Accept-Language", "en-US,en;q=0.8");
+            con.addRequestProperty("User-Agent", "Mozilla");
+            con.addRequestProperty("Referer", "google.com");
+            con.setRequestMethod("POST");
+
+            // POST
+            con.setDoOutput(true);
+            try (OutputStream os = con.getOutputStream()) {
+                os.write(input, 0, input.length);
+            }
+
+            LOG.trace("Redirect to URL: {}", newUrl);
+        }
+
+        BufferedReader in = new BufferedReader(new InputStreamReader(con.getInputStream()));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Use a standard utility function such as Apache Commons IO `IOUtils.toString(InputStream, Charset)` to read data from an input stream. Using a standard utility function makes your code easier to read and less prone to resource leaks. Do not write your own code to read data from the input stream. See the following diff on GitHub that makes this code change: [diff1](https://github.com/MyRobotLab/myrobotlab/commit/a4db462136ebf6c9cb4bc1e60bfac6d6f37a4b6e#diff-a9871318c9e95407fbecd277bb0f4729724c94216d5d5d57d13f743058b9c76dL242)

Learn more: [IOUtils.toString in the Apache Commons IO 2.11.0 API](https://commons.apache.org/proper/commons-io/apidocs/org/apache/commons/io/IOUtils.html#toString-java.io.InputStream-java.nio.charset.Charset-)